File: Python/RemoveNthFromEnd.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def remove_nth_from_end(head: ListNode, n: int) -> ListNode:
    dummy = ListNode(0, head)
    fast = slow = dummy

    # Move fast n+1 steps ahead
    for _ in range(n + 1):
        fast = fast.next

    # Move both pointers until fast reaches end
    while fast:
        fast = fast.next
        slow = slow.next

    # Remove nth node
    slow.next = slow.next.next

    return dummy.next

# Helper to print list
def print_list(head: ListNode):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")

# Example 1: 1->2->3->4->5, n=2
head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
head1 = remove_nth_from_end(head1, 2)
print_list(head1)  # Output: 1 -> 2 -> 3 -> 5 -> None

# Example 2: 5->4->3->2->1, n=5
head2 = ListNode(5, ListNode(4, ListNode(3, ListNode(2, ListNode(1)))))
head2 = remove_nth_from_end(head2, 5)
print_list(head2)  # Output: 4 -> 3 -> 2 -> 1 -> None

File: JAVA/RemoveNthFromEnd.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; this.next = null; }
}

public class RemoveNthFromEnd {

    public static ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;

        // Move fast n+1 steps ahead
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        // Move both pointers until fast reaches end
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // Remove nth node
        slow.next = slow.next.next;

        return dummy.next;
    }

    // Helper to print list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        // Example 1: 1->2->3->4->5, n=2
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        head1.next.next.next = new ListNode(4);
        head1.next.next.next.next = new ListNode(5);

        head1 = removeNthFromEnd(head1, 2);
        printList(head1); // Output: 1 -> 2 -> 3 -> 5 -> null

        // Example 2: 5->4->3->2->1, n=5
        ListNode head2 = new ListNode(5);
        head2.next = new ListNode(4);
        head2.next.next = new ListNode(3);
        head2.next.next.next = new ListNode(2);
        head2.next.next.next.next = new ListNode(1);

        head2 = removeNthFromEnd(head2, 5);
        printList(head2); // Output: 4 -> 3 -> 2 -> 1 -> null
    }
}
